l5kit/l5kit/rasterization/semantic_rasterizer.py

```diff
@@ -80,51 +60,157 @@ def __init__(
-        ego_offset_meters = rot_m @ ego_offset_meters
-
-        sem_im = render_semantic_map(
-            self.semantic_map,
-            x + ego_offset_meters[0],
-            y + ego_offset_meters[1],
-            ego_yaw,
-            raster_size=self.raster_size,
-            pixel_size=self.pixel_size,
+            ego_yaw = agent["yaw"]
+
+        world_to_image_space = world_to_image_pixels_matrix(
+            self.raster_size, self.pixel_size, ego_translation, ego_yaw, self.ego_center,
         )
 
+        # get XY of center pixel in world coordinates
+        center_pixel = np.asarray(self.raster_size) * (0.5, 0.5)
+        center_world = transform_point(center_pixel, np.linalg.inv(world_to_image_space))
+
+        sem_im = self.render_semantic_map(center_world, world_to_image_space, history_tl_faces[0])
         return sem_im.astype(np.float32) / 255
 
+    def render_semantic_map(
+        self, center_world: np.ndarray, world_to_image_space: np.ndarray, tl_faces: np.ndarray
+    ) -> np.ndarray:
+        """Renders the semantic map at given x,y coordinates.
+
+        Args:
+            center_world (np.ndarray): XY of the image center in world ref system
+            world_to_image_space (np.ndarray):
+        Returns:
+            np.ndarray: RGB raster
+
+        """
+
+        img = 255 * np.ones(shape=(self.raster_size[1], self.raster_size[0], 3), dtype=np.uint8)
+
+        # filter using half a radius from the center
+        raster_radius = float(np.linalg.norm(self.raster_size * self.pixel_size)) / 2
+
+        # get active traffic light faces
+        active_tl_ids = set(get_tl_faces_active(tl_faces)["gid"].tolist())
+
+        # plot lanes
+        lanes_lines = defaultdict(list)
+
+        for idx in elements_within_bounds(center_world, self.bounds_info["lanes"]["bounds"], raster_radius):
+            lane = self.proto_API[self.bounds_info["lanes"]["ids"][idx]].element.lane
+
+            # get image coords
+            lane_coords = self.proto_API.get_lane_coords(self.bounds_info["lanes"]["ids"][idx])
+            xy_left = cv2_subpixel(transform_points(lane_coords["xyz_left"][:, :2], world_to_image_space))
+            xy_right = cv2_subpixel(transform_points(lane_coords["xyz_right"][:, :2], world_to_image_space))
+            lanes_area = np.vstack((xy_left, np.flip(xy_right, 0)))  # start->end left then end->start right
+
+            # Note(lberg): this called on all polygons skips some of them, don't know why
+            cv2.fillPoly(img, [lanes_area], (17, 17, 31), lineType=cv2.LINE_AA, shift=CV2_SHIFT)
+
+            lane_type = "default"  # no traffic light face is controlling this lane
+            # TODO should we change the next lane (lanes_ahead) status?
+            lane_tl_ids = set([MapAPI.id_as_str(la_tc) for la_tc in lane.traffic_controls])
+            for tl_id in lane_tl_ids.intersection(active_tl_ids):
```

Comment:
How exactly does this work?
I assume, e.g. a traffic light for going straight and one for turning left would be responsible for different lanes - so the set has len > 0 only if there are multiple traffic light bulbs for the same lane?
Normally, they all would show the same color, but it might be the case that one is broken - in this case we are pessimistic and look for red first etc?

Reply:
A lane has a set of traffic control elements associated. If it is affected by a traffic light, all traffic light bulbs (faces) should be registered in this set. We compute the intersection with the set we have in the zarr (which have the state associated among active, inactive and unknown) and renderise accordingly